### PR TITLE
Added sawmill micro benchmark via jmh

### PR DESCRIPTION
**Pull request by barakm**
Added a set of micro-benchmarks tests for comparing performance of individual pipelines and processors.

### CODE REVIEW

**Comment by barakm** [comment]
Hmm. Looks like running mvn:test will also run all JMH benchmarks. That was not what I intended.

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        );
+
+        pipelineWithVariableRenameProcessors =
+                createPipeline(
+                        IntStream.range(0, NUMBER_OF_FIELDS).boxed().map(
+                                i -> new ProcessorExecutionStep(
+                                        "rename" + i,
+                                        createRenameProcessor(i, 1, templateService)
+                                )
+                        ).collect(Collectors.toList()).toArray(new ExecutionStep[0])
```
Comment (DanMelman) [inline sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:112]:
```suggestion
                        ).toArray(ExecutionStep[]::new)
```

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        pipelineWithVariableRemoveProcessors =
+                createPipeline(
+                        IntStream.range(0, NUMBER_OF_FIELDS).boxed().map(
+                                i -> new ProcessorExecutionStep("drop" + i,
+                                        new RemoveFieldProcessor(
+                                                Arrays.asList(templateService.createTemplate("FieldFrom" + i))
+                                        )
+                                )
+                        ).collect(Collectors.toList()).toArray(new ExecutionStep[0])
```
Comment (DanMelman) [inline sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:130]:
```suggestion
                        ).toArray(ExecutionStep[]::new)
```

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        LinkedHashMap map = new LinkedHashMap<>();
+        if (objects != null) {
+            for (int i = 0; i < objects.length; i++) {
+                map.put(objects[i], objects[++i]);
+            }
+        }
+
+        return new Doc(map);
+    }
```
Comment (DanMelman) [inline sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:212]: what is this for? not used

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+/**
+ * JMH based micro-benchmark suite used to compare pipeline configurations.
+ * Use this as a baseline for comparing the performance of pipelines and specific processors.
+ * The benchmark can be launched from the main() method in this class.
+ */
+@State(Scope.Thread)
+public class SawmillMicroBenchmark {
+    private static final int NUMBER_OF_FIELDS = 100;
+
+    private Map<String, String> documentTemplate;
```
Comment (DanMelman) [inline sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:39]:
```suggestion
    private Map<String, Object> documentTemplate;
```

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                map.put(objects[i], objects[++i]);
+            }
+        }
+
+        return new Doc(map);
+    }
+
+    public Doc createDocFromTemplate(Map<String, String> original) {
+        return new Doc(new LinkedHashMap<>(original));
+    }
```
Comment (DanMelman) [inline sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:216]:
```suggestion
    public Doc createDocFromTemplate(Map<String, Object> original) {
        return new Doc(original);
    }
```

Comment (barakm) [reply sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:215]: Pipeline modified the original map. Have to create a copy here.

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java [hunk 6]
```diff
@@ -187,30 +187,10 @@ public void benchmarkPipelineWithVariableRemoveProcessors() {
...
-
-        LinkedHashMap map = new LinkedHashMap<>();
-        if (objects != null) {
-            for (int i = 0; i < objects.length; i++) {
-                map.put(objects[i], objects[++i]);
-            }
-        }
-
-        return new Doc(map);
+        return new Pipeline("test-pipeline-id", Arrays.asList(steps), true);
```
Comment (DanMelman) [inline sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:190]: it was "false" before this commit

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java (hunk 5 above)
Comment (DanMelman) [reply sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java]: 🤷

File: sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java (hunk 6 above)
Comment (barakm) [reply sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:190]: Makes more sense as true in this context, but has no real impact either way.

Comment (DanMelman) [reply sawmill-benchmark/src/main/java/io/logz/sawmill/benchmark/SawmillMicroBenchmark.java:190]: np, just wanted to make sure it's on purpose

**Comment by barakm** [comment]
Just discovered something annoying. When you run mvn:test, all of the jmh benchmarks run. So these new benchmarks will add ~4 minutes to the build time. I can't see any easy way to exclude the tests from the default run. @DanMelman - wdyt?

**Pull request merged**